File: 2021/python/day_14.py
from collections import deque, Counter


def window(seq, n=2):
    it = iter(seq)
    win = deque((next(it, None) for _ in range(n)), maxlen=n)
    yield win
    append = win.append
    for e in it:
        append(e)
        yield win


def part_1():
    template, insertion_rules = "", {}
    with open("inputs/day14", "r") as file:
        for line in file.read().splitlines():
            if " -> " in line:
                match, insert = line.split(" -> ")
                insertion_rules[match] = insert
            elif line != "":
                template = line

    pairs, chars = Counter(x + y for (x, y) in window(template)), Counter(template)
    for i in range(10):
        for (c1, c2), count in pairs.copy().items():
            insertion = insertion_rules.get(c1 + c2)
            pairs[c1 + c2] -= count
            pairs[c1 + insertion] += count
            pairs[insertion + c2] += count
            chars[insertion] += count

    c = chars.most_common()
    return c[0][1] - c[-1][1]


def part_2():
    template, insertion_rules = "", {}
    with open("inputs/day14", "r") as file:
        for line in file.read().splitlines():
            if " -> " in line:
                match, insert = line.split(" -> ")
                insertion_rules[match] = insert
            elif line != "":
                template = line

    pairs, chars = Counter(x + y for (x, y) in window(template)), Counter(template)
    for i in range(40):
        for (c1, c2), count in pairs.copy().items():
            insertion = insertion_rules.get(c1 + c2)
            pairs[c1 + c2] -= count
            pairs[c1 + insertion] += count
            pairs[insertion + c2] += count
            chars[insertion] += count

    c = chars.most_common()
    return c[0][1] - c[-1][1]


print(part_1())
print(part_2())

File: 2021/src/day_14.rs
use std::{collections::HashMap, fs};

pub const TITLE: &str = "--- Day 14: Extended Polymerization ---";

fn steps(steps: i64, template: String, rules: HashMap<String, String>) -> i64 {
    let mut pairs: HashMap<String, i64> = template.chars().zip(template.chars().skip(1)).fold(
        HashMap::new(),
        |mut counter, (x, y)| {
            *counter.entry(x.to_string() + &y.to_string()).or_insert(0) += 1;
            counter
        },
    );

    let mut chars: HashMap<String, i64> =
        template.chars().fold(HashMap::new(), |mut counter, x| {
            *counter.entry(x.to_string()).or_insert(0) += 1;
            counter
        });

    for _ in 0..steps {
        for (pair, count) in &pairs.clone() {
            match rules.get(pair) {
                Some(to_insert) => {
                    let split = pair
                        .chars()
                        .map(|char| char.to_string())
                        .collect::<Vec<String>>();

                    *pairs.get_mut(pair).unwrap() -= count;
                    *pairs.entry(split[0].clone() + to_insert).or_insert(0) += count;
                    *pairs.entry(to_insert.clone() + &split[1]).or_insert(0) += count;

                    *chars.entry(to_insert.clone()).or_insert(0) += count;
                }
                None => (),
            }
        }
    }

    chars.values().max().unwrap() - chars.values().min().unwrap()
}

pub fn part_1() -> i64 {
    let input: Vec<String> = fs::read_to_string("inputs/day14")
        .expect("Unable to read input")
        .lines()
        .map(|x| x.to_string())
        .collect();

    let rules: HashMap<String, String> = input.iter().filter(|line| line.contains(" -> ")).fold(
        HashMap::new(),
        |mut rules, line| {
            let mut split = line.split(" -> ");
            let parent = split.next().unwrap();
            let child = split.next().unwrap();

            rules.insert(parent.to_string(), child.to_string());
            rules
        },
    );

    steps(10, input[0].clone(), rules)
}

pub fn part_2() -> i64 {
    let input: Vec<String> = fs::read_to_string("inputs/day14")
        .expect("Unable to read input")
        .lines()
        .map(|x| x.to_string())
        .collect();

    let rules: HashMap<String, String> = input.iter().filter(|line| line.contains(" -> ")).fold(
        HashMap::new(),
        |mut rules, line| {
            let mut split = line.split(" -> ");
            let parent = split.next().unwrap();
            let child = split.next().unwrap();

            rules.insert(parent.to_string(), child.to_string());
            rules
        },
    );

    steps(40, input[0].clone(), rules)
}

#[test]
fn test_part_1() {
    assert_eq!(part_1(), 3555);
}

#[test]
fn test_part_2() {
    assert_eq!(part_2(), 4439442043739);
}
